perma_web/perma/utils.py

```diff
@@ -498,26 +504,208 @@ def decrypt_from_perma_payments(ciphertext, encoder=encoding.Base64Encoder):
-        warc_size.append(out.tell())
-        out.seek(0)
-        storages[settings.WARC_STORAGE].store_file(out, destination, overwrite=True)
-        out.close()
+    # the timestamps here are from Link's creation_timestamp, and have "+00:00" at the end, so
+    ts_string = timestamp.isoformat().partition("+")[0] + "Z"
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        # prepare WARC...
+        warc_file = f"{tmpdir}/data.warc.gz"
+        warc = open(warc_file, 'ab+')
+        write_perma_warc_header(warc, guid, timestamp)
+
+        uploaded_file.file.seek(0)
+        write_resource_record_from_asset(uploaded_file.file.read(), warc_url, mime_type, ts_string, warc)
+
+        # create provenance summary and add it to the WARC
+        provenance = loader.get_template("provenance-summary.html")
+        context = {"url": url, "now": ts_string}
+        write_resource_record_from_asset(
+            provenance.render(context).encode(),
+            "file:///provenance-summary.html",
+            "text/html",
+            ts_string,
+            warc
+        )
+        warc.close()
+
+        # ...set up pages.jsonl...
+        pages = [
+            {"format": "json-pages-1.0", "id": "pages", "title": "All Pages"},
+            {
+                "id": f"{uuid.uuid4()}",
+                "url": warc_url,
+                "title": f"User-uploaded file replacing failed capture of {url}",
+                "ts": ts_string
+            },
+            {
+                "id": f"{uuid.uuid4()}",
+                "url": "file:///provenance-summary.html",
+                "title": "Provenance Summary",
+                "ts": ts_string
+            }
+        ]
+        pages_bytes = "\n".join([json.dumps(page) for page in pages]).encode("utf-8")
+
+        # ...CDXJ index...
+        targets = ["file:///provenance-summary.html", warc_url]
+        selected_warc_headers = parse_warc(warc_file, warc_url)
+        cdxj = {
+            target: json.dumps({
+                "url": target,
+                "mime": "text/html" if target.endswith(".html") else mime_type,
+                "status": 200,
+                "digest": selected_warc_headers[target]["digest"],
+                "length": selected_warc_headers[target]["length"],
+                "offset": selected_warc_headers[target]["offset"],
+                "filename":"data.warc.gz"
+            }).replace(" ", "") for target in targets
+        }
+        ts = timestamp.strftime("%Y%m%d%H%M%S")
+        index = "\n".join(
+            [
+                f"{target} {ts} {cdxj[target]}"
+                for target in targets
+            ]
+        )
+
+        # ...datapackage...
+        datapackage = {
+            "profile": "data-package",
+            "wacz_version": "1.1.1",
+            "title": title,
+            "description": f"User upload for {url}",
+            "mainPageURL": warc_url,
+            "created": ts_string,
+            "software": "Perma.cc",  # version?
```

Comment:
`# version?`

Hmmm... 

So if we wanted to consider the production "tag" the version, then we'd have to somehow expose that to the application and interpolate here... Which sounds like a bit of a pain in the butt? (Some salt state that runs a command that gets the current tag and puts it into an ENV var, and then an application setting that reads that ENV var?)

Some new version that we just increment if/when we change this code (we'll be prone to forget, I imagine).

Or skip entirely?

Hmmmmmmmmm.

Reply:
Yeah, the automated mechanism for accomplishing this would be slightly baroque and the manual way would be failure-prone. I think I still want to add it? Let me think about the best approach -- I think it will be something like what you describe, maybe writing settings directly rather than using an env var.

Reply:
If we do set this up, should the setting be visible in templates? Would we want to expose it elsewhere in the application?

Reply:
I don't know why I feel like we did this once before, or something like it, and it was exposed in the templates, and it was read by an LIL service or uptimerobot or something....

Reply:
Here's how we did it before: https://github.com/harvard-lil/perma/commit/7e0a04b0403e9d44b95ce411fd6e6b86216099d1

Reply:
I now have a mechanism for interpolation in deployment; I don't love it, but it works, and I may revisit it later.